### ISSUE TITLE
Unit testing exported projects

### ISSUE THREAD

**Issue by cryptowyrm** [oakes/Lightmod#7]
Lightmod is the best thing since sliced bread, but I can’t figure out how to add basic unit tests to the applications exported from it.

I tried adding boot-cljs-test to the „basic“ template output with a little help from the modern-cljs tutorial which covers boot-cljs-test in chapter 15, but I wasn't able to figure out how to modify the „boot run“ task of the Lightmod template to make it work.

Could unit testing be added to the template, so that users of Lightmod would just have to add a test file like test/projectname/client_test.cljs to make it work?

If not because you want to keep it as simple as possible, I’d really appreciate it if you could provide instructions on how to do it, or point me in the right direction.

**Comment by cryptowyrm** [comment]
Finally figured it out, but it took me about ten hours so I'll leave this issue open because I think it would be really helpful for people new to ClojureScript if this was added to Lightmod's boot template at some point if possible, together with testing Clojure code but I don't use that right now so I haven't looked into that personally.

**Here's what I changed to make CLJS tests work:**

The way modern-cljs explains configuring boot for testing ClojureScript is basically not a good fit for the setup Lightmod projects use, that's why I had so much trouble with it. Instead of doing everything in the "boot run" task, it's better to have a separate "boot auto-test" task and then start both in two separate terminal sessions.

Reagent's render method also needs to be put into a function so that it isn't run when the code is tested, phatomjs will complain otherwise. I put it into a function called init and then added projectname.client/init to the :init-fns parameter in both dev-resources and prod-resources .edn files 

I created a new test-resources/projectname/main.cljs.edn file (without setting :init-fns):

    {:require [projectname.client], :init-fns [], :compiler-options {}}

And added three boot tasks to build.boot:

    (deftask testing []
      (merge-env! :source-paths #{"test"})
      identity)

    (deftask test-once []
      (set-env! :resource-paths #{"resources" "test-resources"})
      (comp
        (testing)
        (test-cljs :js-env :phantom)))

    (deftask auto-test []
      (set-env! :resource-paths #{"resources" "test-resources"})
      (comp
        (testing)
        (watch)
        (test-cljs :js-env :phantom)))

Then a test can be written in test/projectname/client_test.cljs:

    (ns projectname.client-test
      (:require [projectname.client :as app]
                [cljs.test :refer-macros [deftest testing is are]]))

    (deftest test-testing
      (testing "Making sure tests work"
        (is (= 2 2))))
